Shadow Stack and Indirect Branch Tracking support

Would it be possible to add support for either of these features to libunwind [1]?

Compilers will generate code supporting these features when -fcf-protection option is used.  Assembly code needs to add a .note.gnu.property section specifying which features are supported which serves a similar function to .note.GNU-stack and is checked by the linker.  I do not believe any other changes would be needed for shadow stack support unless the library makes use of the stack.  Indirect branch tracking requires all indirect branch targets be marked with ENDBR.  You can see how support has been added to a number of projects at [2].  Shadow stack support has been merged into the kernel for 6.6 [3] and I expect compatible glibc support for it will be added once 6.6 is released.

[1]: https://www.intel.com/content/dam/develop/external/us/en/documents/catc17-introduction-intel-cet-844137.pdf
[2]: https://gitlab.com/cet-software
[3]: https://github.com/torvalds/linux/commit/df57721f9a63e8a1fb9b9b2e70de4aa4c7e0cd2e

@eslerm it has been partially resolved; all the libraries apart from `libunwind-setjmp.so` are now built with shadow stack and indirect branch tracking enabled.
Steps I used to check the current git head b2d2e81e5fe53b01a0dd39e2e634f963ebeaab43:
```
$ git clone https://github.com/libunwind/libunwind.git
$ cd libunwind
$   git describe --tags
v1.8-branchpoint-126-gb2d2e81e
$ autoreconf -fiv
$ LDFLAGS='-Wl,-z,cet-report=error' CFLAGS='-fcf-protection' ./configure
$ make
....
/usr/bin/ld: x86_64/.libs/longjmp.o: error: missing IBT and SHSTK properties
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:4110: libunwind-setjmp.la] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: Leaving directory '/build/libunwind/src/libunwind/src'
make[1]: *** [Makefile:2970: all] Error 2
make[1]: Leaving directory '/build/libunwind/src/libunwind/src'
make: *** [Makefile:631: all-recursive] Error 1
```
Full build output attached.
[libunwind-v1.8.branchpoint.r126.gb2d2e81-1-x86_64-build.log](https://github.com/user-attachments/files/21631399/libunwind-v1.8.branchpoint.r126.gb2d2e81-1-x86_64-build.log)

--- a/src/x86_64/getcontext.S
+++ b/src/x86_64/getcontext.S
@@ -38,9 +38,7 @@ WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.  */
 	.type _Ux86_64_getcontext, @function
 _Ux86_64_getcontext:
 	.cfi_startproc
-#if defined(__CET__) && __CET__ & 0x01
-	endbr64
-#endif
+	_CET_ENDBR
 
 	/* Callee saved: RBX, RBP, R12-R15  */
 	movq %r12, UC_MCONTEXT_GREGS_R12(%rdi)
@@ -141,23 +139,3 @@ _Ux86_64_getcontext_trace:
 
       /* We do not need executable stack.  */
       .section        .note.GNU-stack,"",@progbits
-
-#if defined(__CET__)
-      /* ELF note indicating control-flow protection in effect */
-	.section NOTE_GNU_PROPERTY_SECTION_NAME,"a"
-	.align 8
-	.long 1f - 0f
-	.long 4f - 1f
-	.long NT_GNU_PROPERTY_TYPE_0
-0:
-	.string "GNU"
-1:
-	.align 8
-	.long GNU_PROPERTY_X86_FEATURE_1_AND_LO
-	.long 3f - 2f
-2:
-	.long __CET__ 
-3:
-	.align 8
-4:
-#endif

--- a/src/x86_64/longjmp.S
+++ b/src/x86_64/longjmp.S
@@ -23,9 +23,12 @@ LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
 WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.  */
 
+#include "ucontext_i.h"
+
 	.globl _UI_longjmp_cont
 	.type _UI_longjmp_cont, @function
 _UI_longjmp_cont:
+	_CET_ENDBR
 	push %rax		/* push target IP as return address */
 	mov %rdx, %rax		/* set up return-value */
 	retq

--- a/src/x86_64/setcontext.S
+++ b/src/x86_64/setcontext.S
@@ -36,9 +36,7 @@ WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.  */
 	.type _Ux86_64_setcontext, @function
 _Ux86_64_setcontext:
 	.cfi_startproc
-#if defined(__CET__) && __CET__ & 0x01
-	endbr64
-#endif
+	_CET_ENDBR
 
 #if defined __linux__ || defined __sun
         /* restore fp state */
@@ -90,24 +88,3 @@ _Ux86_64_setcontext:
 
       /* We do not need executable stack.  */
       .section        .note.GNU-stack,"",@progbits
-
-#if defined(__CET__)
-      /* ELF note indicating control-flow protection in effect */
-	.section NOTE_GNU_PROPERTY_SECTION_NAME,"a"
-	.align 8
-	.long 1f - 0f
-	.long 4f - 1f
-	.long NT_GNU_PROPERTY_TYPE_0
-0:
-	.string "GNU"
-1:
-	.align 8
-	.long GNU_PROPERTY_X86_FEATURE_1_AND_LO
-	.long 3f - 2f
-2:
-	.long __CET__ 
-3:
-	.align 8
-4:
-#endif
-

--- a/src/x86_64/siglongjmp.S
+++ b/src/x86_64/siglongjmp.S
@@ -32,9 +32,7 @@ WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.  */
 	.globl _UI_siglongjmp_cont
 	.type _UI_siglongjmp_cont, @function
 _UI_siglongjmp_cont:
-#if defined(__CET__) && __CET__ & 0x01
-	endbr64
-#endif
+	_CET_ENDBR
 #if defined(__linux__) || defined(__QNX__)
 	retq
 #elif defined(__FreeBSD__)
@@ -53,23 +51,3 @@ _UI_siglongjmp_cont:
 	.size _UI_siglongjmp_cont, . - _UI_siglongjmp_cont
 	/* We do not need executable stack.  */
 	.section	.note.GNU-stack,"",@progbits
-#if defined(__CET__)
-	/* ELF note indicating control-flow protection in effect */
-	.section NOTE_GNU_PROPERTY_SECTION_NAME,"a"
-	.align 8
-	.long 1f - 0f
-	.long 4f - 1f
-	.long NT_GNU_PROPERTY_TYPE_0
-0:
-	.string "GNU"
-1:
-	.align 8
-	.long GNU_PROPERTY_X86_FEATURE_1_AND_LO
-	.long 3f - 2f
-2:
-	.long __CET__ 
-3:
-	.align 8
-4:
-#endif
-

--- a/src/x86_64/ucontext_i.h
+++ b/src/x86_64/ucontext_i.h
@@ -122,17 +122,10 @@ WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.  */
 
 #endif
 
-/**
- * Details for the GNU note.gnu.property extension.
- *
- * These are required for indicating the presence of the Intel Indirect Branch
- * Tracking and Shadow Stack features (the Intel CET extension).
- *
- * https://software.intel.com/en-us/articles/intel-sdm
- *
- */
-#define NOTE_GNU_PROPERTY_SECTION_NAME ".note.gnu.property"
-#define NT_GNU_PROPERTY_TYPE_0                        5
-#define GNU_PROPERTY_X86_FEATURE_1_AND_LO    0xc0000002
+#if defined(__CET__)
+# include <cet.h>
+#else
+# define _CET_ENDBR
+#endif
 
 #endif /* unw_ucontext_i_h */

--- a/tests/x64-test-dwarf-expressions.S
+++ b/tests/x64-test-dwarf-expressions.S
@@ -33,6 +33,12 @@
 # these registers and cause this test to fail mysteriously.
 
 
+#if defined(__CET__)
+# include <cet.h>
+#else
+# define _CET_ENDBR
+#endif
+
 ########################################################
 # Test: Restoring a register using a DW_CFA_expression #
 # which uses implicit CFA pushed onto stack.           #
@@ -41,6 +47,7 @@
 .type DW_CFA_expression_testcase STT_FUNC
 DW_CFA_expression_testcase:
   .cfi_startproc
+  _CET_ENDBR
   push %r12
   .cfi_adjust_cfa_offset 8
   # Move our sentinel (known) value into non-volatile (Callee-saved) r12

